File: InfiniteScrollViewDemo/StoryModel.m
//
//  StoryModel.m
//  InfiniteScrollViewDemo
//
//  Created by pronebird on 10/05/14.
//  Copyright (c) 2014 codeispoetry.ru. All rights reserved.
//

#import "StoryModel.h"

static NSString * const StoryModelAttributeTitle  = @"title";
static NSString * const StoryModelAttributeAuthor = @"author";
static NSString * const StoryModelAttributeURL    = @"url";

@implementation StoryModel

+ (instancetype)modelWithDictionary:(NSDictionary *)dictionary {
    return [[self alloc] initWithDictionary:dictionary];
}

+ (NSArray<StoryModel *> *)modelsFromArray:(NSArray<NSDictionary *> *)array {
    NSMutableArray *models = [[NSMutableArray alloc] init];
    
    for(NSDictionary *dict in array) {
        StoryModel *model = [self modelWithDictionary:dict];
        if(model) {
            [models addObject:model];
        }
    }
    
    return [models copy];
}

- (id)initWithDictionary:(NSDictionary *)dictionary {
    self = [super init];
    if(!self) {
        return nil;
    }
    
    self.title = dictionary[StoryModelAttributeTitle];
    self.author = dictionary[StoryModelAttributeAuthor];
    
    // sometimes HN returns some trash
    NSString *url = dictionary[StoryModelAttributeURL];
    
    if(!url || [[NSNull null] isEqual:url]) {
        return nil;
    }
    
    self.url = [NSURL URLWithString:url];
    
    return self;
}

@end

File: InfiniteScrollViewDemoSwift/StoryModel.swift
//
//  StoryModel.swift
//  InfiniteScrollViewDemoSwift
//
//  Created by pronebird on 5/3/15.
//  Copyright (c) 2015 pronebird. All rights reserved.
//

import Foundation

private struct StoryModelAttributes {
    static let url = "url"
    static let title = "title"
    static let author = "author"
}

class StoryModel: NSObject {
    var title: String?
    var author: String?
    var url: URL
    
    init?(_ dictionary: [String: AnyObject]) {
        // sometimes HN returns some trash
        guard let urlString = dictionary[StoryModelAttributes.url] as? String,
              let urlObject = URL(string: urlString)
        else {
            return nil
        }
    
        title = dictionary[StoryModelAttributes.title] as? String
        author = dictionary[StoryModelAttributes.author] as? String
        url = urlObject
        
        super.init()
    }

}
